src/core/qgsjsonutils.cpp

```diff
@@ -278,6 +295,15 @@ QString QgsJSONUtils::exportAttributes( const QgsFeature &feature )
       attrs += QLatin1String( ",\n" );
 
     QVariant val = feature.attributes().at( i );
+
+    if ( layer )
+    {
+      QgsEditorWidgetSetup setup = layer->fields().at( i ).editorWidgetSetup();
+      QgsFieldFormatter *fieldFormatter = QgsApplication::fieldFormatterRegistry()->fieldFormatter( setup.type() );
+      if ( fieldFormatter != QgsApplication::fieldFormatterRegistry()->fallbackFieldFormatter() )
```

Comment:
The fallback formatter converts non string variants to string. We need to keep them as numeric data types where possible